### PR TITLE
Helm chart for kubernetes deployment

### LINKED ISSUE #26: Official helm chart to deploy on kubernetes and integrate with prometheus-operator
I have started work on it and will raise a PR soon. Testing in progress

### PR DESCRIPTION
**Pull request by mikhailadvani**
Once merged, we can use github pages to publish the chart and update the documentation.

Closes #26

### CODE REVIEW

**Comment by Strech** [comment]
@mikhailadvani Hey 👋 Thanks for your contrubision. I hope I will have enough time to make a review this weekend.

**Comment by mikhailadvani** [comment]
@Strech I have added a few enhancements that I could think of related to security & reliability

File: helm-charts/sidekiq-prometheus-exporter/values.yaml [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+tolerations: []
+
+affinity: {}
+
+externalEnvVars:
+  name: ""
+  type: "ConfigMap"
+
+envVars: {}
+  # REDIS_HOST: "localhost"
```
[the remark this answers is not in the document]
Comment (mikhailadvani) [reply helm-charts/sidekiq-prometheus-exporter/values.yaml:43]: I have added the remaining default values in the values file in comments. the generic envVars approach allows additional/different environment variables as someone might want to use with a custom docker image. hence the all caps approach

File: helm-charts/sidekiq-prometheus-exporter/values.yaml [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+serviceAccount:
+  # Specifies whether a ServiceAccount should be created
+  create: true
+  # The name of the ServiceAccount to use.
+  # If not set and create is true, a name is generated using the fullname template
+  name:
+  # annotations:
+  # Will add the provided map to the annotations for the created serviceAccount
+
+rbac:
```
[the remark this answers is not in the document]
Comment (mikhailadvani) [reply helm-charts/sidekiq-prometheus-exporter/values.yaml:66]: updated based on https://helm.sh/docs/chart_best_practices/rbac/#using-rbac-resources

File: helm-charts/sidekiq-prometheus-exporter/values.yaml [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  interval: 10s
+  labels: {}
+  scrapeTimeout: 5s
+
+serviceAccount:
+  # Specifies whether a ServiceAccount should be created
+  create: true
+  # The name of the ServiceAccount to use.
+  # If not set and create is true, a name is generated using the fullname template
+  name:
```
[the remark this answers is not in the document]
Comment (mikhailadvani) [reply helm-charts/sidekiq-prometheus-exporter/values.yaml:64]: updated based on https://helm.sh/docs/chart_best_practices/rbac/#using-rbac-resources

File: helm-charts/sidekiq-prometheus-exporter/values.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (Strech) [reply helm-charts/sidekiq-prometheus-exporter/values.yaml:43]: Yeah, I totally agree with caps and I think it's right, but my comment was about where to write docs, they should be placed on top of variable, not bottom. Bottom goes anything underneath to clarify what you can nest in the higher level key.

Here is an example https://github.com/bitnami/charts/blob/master/bitnami/redis/values.yaml

**Comment by Strech** [comment]
@mikhailadvani I have some additional feedback. I think the best for the user will be an ability to spin up this helm with just a few values set, the ones which are crucial to the application.

I just tried to run the latest PR state in a console and get that:

```bash
$ helm install --generate-name --dry-run helm-charts/sidekiq-prometheus-exporter
Error: unable to build kubernetes objects from release manifest: unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1"
```

**Comment by mikhailadvani** [comment]
@Strech you get the error because you do not have the prometheus-operator installed. I have gone in for the prometheus-operator approach rather than the vanilla prometheus approach since it allows easier wiring in through service-discovery labels

File: helm-charts/sidekiq-prometheus-exporter/templates/service.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ include "sidekiq-prometheus-exporter.fullname" . }}
+  labels:
+{{ include "sidekiq-prometheus-exporter.labels" . | indent 4 }}
```
[the remark this answers is not in the document]
Comment (Strech) [reply helm-charts/sidekiq-prometheus-exporter/templates/service.yaml:6]: The issue is nesting, it's not visual and not obvious. I can give an example with `Deployment` which is 100% same with Service

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: {{ include "sidekiq-prometheus-exporter.fullname" . }}
  labels:
    {{- include "sidekiq-prometheus-exporter.labels" . | nindent 4 }}
```

File: helm-charts/sidekiq-prometheus-exporter/values.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (mikhailadvani) [reply helm-charts/sidekiq-prometheus-exporter/values.yaml:43]: it is supposed to be nested under envVars to allow dynamic usage through [this](https://github.com/mikhailadvani/sidekiq-prometheus-exporter/blob/392d9c0b896c534512dd2ad26792acb1ad373c60/helm-charts/sidekiq-prometheus-exporter/templates/deployment.yaml#L27-L32). I am not sure whether I understood what you meant in the redis example that you shared

File: helm-charts/sidekiq-prometheus-exporter/templates/service.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (mikhailadvani) [reply helm-charts/sidekiq-prometheus-exporter/templates/service.yaml]: done

File: helm-charts/sidekiq-prometheus-exporter/values.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (Strech) [reply helm-charts/sidekiq-prometheus-exporter/values.yaml:43]: Ah, I got what you mean, yep, all good here

**Comment by Strech** [comment]
> @Strech you get the error because you do not have the prometheus-operator installed. I have gone in for the prometheus-operator approach rather than the vanilla prometheus approach since it allows easier wiring in through service-discovery labels

Yes, you are right. I disable `serviceMonitor` and was able to dry-run it. But I mean in general, that I think that as less you need entities to spin the simple exporter as better. And all non-essential for application entities, in my opinion, should be possible to enable but disabled by default.

And again, if you turn them on – less configuration as possible to make them running.

**Comment by mikhailadvani** [comment]
switched to serviceMonitor.enabled false and added an empty podAnnotations map for the ability to prometheus to scrape metrics in a non-operator setup

**Comment by mikhailadvani** [comment]
addressed points 1 & 3. I do not agree on 2. keeping the service port as 80 with cluster IP keeps it simple for anyone consuming the service (rather than the pod) over http within the cluster as http://<service-name>. the port is an internal configuration of the setup and IMHO a candidate for abstraction. I look at the service as a load balancer/reverse proxy abstracting away the internal implementation and would prefer to stick to standard ports. your thoughts?

**Comment by Strech** [comment]
@mikhailadvani your point sounds reasonable. Agree 👍

**Comment by mikhailadvani** [comment]
all other comments have already been addressed

**Comment by Strech** [comment]
@mikhailadvani Hey, I want to say thank you for your great contribution 👏 🎉

**Comment by mikhailadvani** [comment]
you are welcome. more issues and PRs coming soon

**Comment by Strech** [comment]
@mikhailadvani Currently I'm re-shaping cart, so be aware 👀

**Pull request merged**